corehq/apps/enterprise/enterprise.py

```diff
@@ -241,23 +245,56 @@ def headers(self):
     def _query(self, domain_name):
         time_filter = form_es.submitted
 
-        users_filter = form_es.user_id(UserES().domain(domain_name).mobile_users().show_inactive()
-                                    .values_list('_id', flat=True))
+        users_filter = form_es.user_id(
+            UserES().domain(domain_name).mobile_users().show_inactive().size(self.MAXIMUM_USERS_PER_DOMAIN + 1)
+            .values_list('_id', flat=True)
+        )
 
-        query = (form_es.FormES()
-                 .domain(domain_name)
-                 .filter(time_filter(gte=self.datespan.startdate,
-                                     lt=self.datespan.enddate_adjusted))
-                 .filter(users_filter))
+        if len(users_filter) > self.MAXIMUM_USERS_PER_DOMAIN:
+            raise TooMuchRequestedDataError(
+                _('Domain {name} has too many users. Maximum allowed is: {amount}')
+                .format(name=domain_name, amount=self.MAXIMUM_USERS_PER_DOMAIN)
+            )
+
+        query = (
+            form_es.FormES()
+            .domain(domain_name)
+            .filter(time_filter(gte=self.datespan.startdate, lt=self.datespan.enddate_adjusted))
+            .filter(users_filter)
+        )
         return query
 
     def hits(self, domain_name):
-        return self._query(domain_name).run().hits
+        return raise_after_max_elements(
+            self._query(domain_name).scroll(),
+            self.MAXIMUM_ROWS_PER_REQUEST,
+            self._generate_data_error()
+        )
+
+    def _generate_data_error(self):
+        return TooMuchRequestedDataError(
+            _('{name} contains too many rows. Maximum allowed is: {amount}. Please narrow the date range'
+                ' to fetch a smaller amount of data').format(
+                    name=self.account.name, amount=self.MAXIMUM_ROWS_PER_REQUEST)
+        )
+
+    @property
+    def rows(self):
+        total_rows = 0
+        rows = []
+        for domain_obj in self.domains():
+            domain_rows = self.rows_for_domain(domain_obj)
+            total_rows += len(domain_rows)
+            if total_rows > self.MAXIMUM_ROWS_PER_REQUEST:
```

Comment:
With this check here, is it necessary to also check for rows over the limit in `raise_after_max_elements`?  Is that so that it fails earlier if one domain is especially large?

Reply:
Yeah -- we're seeing one specific domain that generates over a million forms each month, so that domain alone could cause performance issues without the domain-specific check